Update LSP6 Specs with new behavior of AllowedCalls (#191)
lukso-network/LIPs: LSPs/LSP-6-KeyManager.md

Since the `valueType` of this data key is `bytes32`, up to 255 different permissions can be defined. This includes the [default permissions](#permissions) defined. Custom permissions can be defined on top of the default one.

#### AddressPermissions:AllowedCalls:\<address\>

```json
{
    "name": "AddressPermissions:AllowedCalls:<address>",
    "key": "0x4b80742de2bf393a64c70000<address>",
    "keyType": "MappingWithGrouping",
    "valueType": "(bytes4,address,bytes4,bytes4)[CompactBytesArray]",
    "valueContent": "(BitArray,Address,Bytes4,Bytes4)"
}
```

> Contains a compact bytes array of operations restrictions, addresses, interfaceIds and function selectors a controller address is allowed to execute and interact with.

Reviewer: I would not use the term "operations restrictions" but instead "calls restrictions" as "TRANSFER VALUE" is not an operation.

We can also keep what was in parentheses.

```suggestion
Contains a compact bytes array of call restrictions, addresses, interfaceIds and function selectors a controller address is allowed to execute and interact with.
```

> Each entry (allowed call) is made of four elements concatenated together as a tuple that forms a final `bytes32` long value.

Reviewer: Maybe we can shorten here by saying _"each allowed call"_ instead of _"each entry (allowed call)"_

The full list of allowed calls MUST be constructed as a [CompactBytesArray](./LSP-2-ERC725YJSONSchema.md#bytescompactbytesarray) according to [LSP2-ERC725YJSONSchema] as follow:

```js
<0020> <bytes4 restrictionOperations> <bytes20 allowedAddress> <bytes4 allowedInterfaceId> <bytes4 allowedFunction> <0020> ... <0020> ...
```

> **NB:** the three dots `...` are placeholders for `<bytes4 restrictionOperations> <bytes20 allowedAddress> <bytes4 allowedInterfaceId> <bytes4 allowedFunction>` and used for brievity.

> - `0020`: **0020** in decimals is **32**, which is the sum of bytes length of the four elements below concatenated together.

Reviewer: ```suggestion
- `0020`: **0020** in decimals is **32**, the total number of bytes of the four elements below concatenated together.
```

> - `restrictionOperations`: A bitArray that represents the list of operations that the restrictions listed (address - interfaceId - function) applies for.

Reviewer: to be reviewed with @frozeman this part on Monday. Seems good to me but I guess we have to agree on the terminology ("restrictionOperations"? "callTypes"? "callRestrictions"?)

>     The operations are defined with specific bits, starting from the 4th byte from the most right: 

Reviewer: Suggesting to use the term "least significant bit"

```suggestion
    The operations are defined with specific bits, starting from the least significant bit the most right: 
```

![image](https://user-images.githubusercontent.com/31145285/225966727-6d3a6daa-cb6d-474e-8582-09edf88c6a60.png)
https://www.techopedia.com/definition/8030/least-significant-bit-lsb#:~:text=In%20computing%2C%20the%20least%20significant,to%20transmission%20of%20binary%20numbers.

    - transferValue `00000001`
    - call `00000010`
    - staticcall `00000100`
    - delegatecall `00001000` 

    Custom implementations of the LSP6 Standard can add more operations as needed.

- `allowedAddress`: The address called by the target contract.
- `allowedInterfaceId`: The ERC165 interface id being supported by the contract called from the target.
- `allowedFunction`: The function selector being called on the contract called by the target contract.

- If the value of the data key is **empty**, execution is disallowed.

> - `restrictionOperations` cannot be discarded. 

Reviewer: Suggesting to use the term MUST NOT here

```suggestion
- `restrictionOperations` MUST NOT be discarded. 
```

- Check is discarded for an element if the value is full `ff` bytes. e.g, `0xffffffff` for interfaceIds and function selectors and `0xffffffffffffffffffffffffffffffffffffffff` for addresses. There MUST be at most 2 discarded checks, meaning `0xffffffffffffffffffffffffffffffffffffffffffffffffffffffff` data key is disallowed. 

**Example 1:**

If address A has [CALL](#permissions) permission, and have the following value for AllowedCalls:

```
0x002000000002cafecafecafecafecafecafecafecafecafecafe11223344bb11bb11
```

> The `restrictionOperations` is **`0x00000002`**, it means that the restrictions above only applies when the operation **[CALL](https://github.com/ERC725Alliance/ERC725/blob/develop/docs/ERC-725.md#execute)** is being executed.

Reviewer: ```suggestion
The `restrictionOperations` is **`0x00000002`**, meaning the restrictions above only applies when the operation **[CALL](https://github.com/ERC725Alliance/ERC725/blob/develop/docs/ERC-725.md#execute)** is being executed.
```

The address A is allowed to interact via the **CALL** operation with the function selector **`0xbb11bb11`** on the **`0xcafecafecafecafecafecafecafecafecafecafe`** address as long as the address supports **`0x11223344`** interfaceId through [ERC165]. 

<br>

**Example 2:**

If address B has [DELEGATECALL](#permissions) permission, and have the following value for AllowedCalls:

```
0x002000000004cafecafecafecafecafecafecafecafecafecafeffffffffffffff002000000004ffffffffffffffffffffffffffffffffffffffff68686868ffffffff
```

The `restrictionOperations` on both elements is **`0x00000004`**, it means that the restrictions above only applies when the operation **[DELEGATECALL](https://github.com/ERC725Alliance/ERC725/blob/develop/docs/ERC-725.md#execute)** is being executed.

> The address B is allowed to interact via **DELEGATECALL** with:

Reviewer: suggested changes:

The address B is allowed to interact with:
- the address **`0xcafecafecafecafecafecafecafecafecafecafe`** via DELEGATECALL without any restriction on the interfaceId or the function selector.
- **any address** supporting the **`0x68686868`** interfaceId without any restriction on the function.

- the address **`0xcafecafecafecafecafecafecafecafecafecafe`** without any restriction on the interfaceId or the function selector.
- **any address** supporting the **`0x68686868`** interfaceId without any restriction on the function.

<br>

**Example 3:**

If address B has [DELEGATECALL](#permissions) permission, and have the following value for AllowedCalls:

```
0x002000000002cafecafecafecafecafecafecafecafecafecafe11223344bb11bb11
```

The `restrictionOperations` is **`0x00000002`**, it means that the restrictions above only applies when the operation **[CALL](https://github.com/ERC725Alliance/ERC725/blob/develop/docs/ERC-725.md#execute)** is being executed.

Given that the address B have the **DELEGATECALL** Permission and the restrictions only applies for the **CALL** operation, any execution of **DELEGATECALL** operation will fail even to the function **`0xbb11bb11`** on **`0xcafecafecafecafecafecafecafecafecafecafe`** address.

<br>

**Example 4:**

If address B has [TRANSFERVALUE](#permissions) and [CALL](#permissions) permissions, and have the following value for AllowedCalls:

```
0x002000000003cafecafecafecafecafecafecafecafecafecafe11223344bb11bb11
```

> The `restrictionOperations` is **`0x00000003`**, it means that the restrictions above applies when the operation **[CALL](https://github.com/ERC725Alliance/ERC725/blob/develop/docs/ERC-725.md#execute)** or **[TRANSFERVALUE](https://github.com/ERC725Alliance/ERC725/blob/develop/docs/ERC-725.md#execute)** is being executed.

Reviewer: should we call this `callTypes`?

> The address B is allowed to **CALL** or **TransferValue** to:

Reviewer: ```suggestion
The controller B is allowed to **CALL** or **TransferValue** to:
```

> - the function selector **`0xbb11bb11`** on the **`0xcafecafecafecafecafecafecafecafecafecafe`** address as long as the address supports **`0x11223344`** interfaceId through [ERC165].

Reviewer: CALL with value 
or
CALL without value

Reviewer: it is about how to formulate it here (there is no perfect way)

controller is allowed to:

- _interact with the ERC20 contract (ERC20 interface ID `0xblabla...`)_
- deployed at address `0xcafecafe...`
- but call only the function `0xbb11bb11` on it)

The controller can call this function on this ERC20 contract either with or without transferring value.

<br>

**Example 5:**

If address B has [TRANSFERVALUE](#permissions) and [CALL](#permissions) permissions, and have the following value for AllowedCalls:

```
0x002000000001cafecafecafecafecafecafecafecafecafecafe11223344bb11bb11002000000002ffffffffffffffffffffffffffffffffffffffff68686868ffffffff
```

Each element of the compact bytes array have a different `restrictionOperations`:

- **`0x000000001`** for the first element allowing the address B to only **TransferValue** to the function **`0xbb11bb11`** on **`0xcafecafecafecafecafecafecafecafecafecafe`** address as long as the address supports **`0x11223344`** interfaceId through [ERC165].

> - **`0x000000002`** for the second element allowing the address B to only **Call** any function on any address as long as the address supports **`0x68686868`** interfaceId through [ERC165].

Reviewer: ```suggestion
- **`0x000000002`** for the second element allowing the address B to only **Call** any function on any contract that support **`0x68686868`** interfaceId [ERC165].
```
